File: javaNode/src/main/java/org/kevoree/library/wrapper/WrapperFactory.java
package org.kevoree.library.wrapper;

import org.kevoree.*;
import org.kevoree.api.ChannelContext;
import org.kevoree.api.ChannelContextImpl;
import org.kevoree.api.ModelService;
import org.kevoree.api.RuntimeService;
import org.kevoree.library.compare.KevoreeThreadGroup;
import org.kevoree.modeling.api.KMFContainer;

/**
 *
 * Created by duke on 9/26/14.
 */
public class WrapperFactory {

    public WrapperFactory(String nodeName) {
        this.nodeName = nodeName;
    }

    private String nodeName;

    public KInstanceWrapper wrap(KMFContainer instance, Object instanceObj, RuntimeService runtimeService,
                                 ModelService modelService)
            throws IllegalArgumentException {

        KInstanceWrapper wrapper = null;
        if (instance instanceof ComponentInstance) {
            wrapper = new ComponentWrapper();
        }
        if (instance instanceof Group) {
            wrapper = new GroupWrapper();
        }
        if (instance instanceof Channel) {
            wrapper = new ChannelWrapper((ChannelContextImpl) runtimeService.getService(ChannelContext.class));
        }
        if (instance instanceof ContainerNode) {
            wrapper = new NodeWrapper();
        }
        if (wrapper != null) {
            wrapper.setModelService(modelService);
            wrapper.setRuntimeService(runtimeService);
            wrapper.setStarted(false);
            wrapper.setThreadGroup(new KevoreeThreadGroup(instance.path()));
            wrapper.setNodeName(nodeName);
            wrapper.setTargetObj(instanceObj);
            wrapper.setModelElement((Instance) instance);
            return wrapper;
        } else {
            throw new IllegalArgumentException("Instance type must be ComponentInstance, Group, Channel or ContainerNode (current: " + instance.metaClassName() + ")");
        }
    }

}

File: java/org.kevoree.library.java.javaNode/src/main/java/org/kevoree/library/defaultNodeTypes/wrapper/WrapperFactory.kt
package org.kevoree.library.defaultNodeTypes.wrapper

import org.kevoree.modeling.api.KMFContainer
import org.kevoree.ComponentInstance
import org.kevoree.Group
import org.kevoree.Channel
import org.kevoree.ContainerNode
import org.kevoree.api.BootstrapService

/**
 * Created with IntelliJ IDEA.
 * User: duke
 * Date: 03/12/2013
 * Time: 09:33
 */

open class WrapperFactory(val nodeName: String) {

    open fun wrap(modelElement: KMFContainer, newBeanInstance: Any, tg: ThreadGroup, bs: BootstrapService): KInstanceWrapper {
        when(modelElement) {
            is ComponentInstance -> {
                return ComponentWrapper(modelElement, newBeanInstance, nodeName, tg, bs)
            }
            is Group -> {
                return GroupWrapper(modelElement,newBeanInstance, nodeName, tg, bs)
            }
            is Channel -> {
                return ChannelWrapper(modelElement, newBeanInstance, nodeName, tg, bs)
            }
            is ContainerNode -> {
                return NodeWrapper(modelElement, newBeanInstance, nodeName, tg, bs)
            }
            else -> {
                throw Exception("Unknow instance type " + modelElement.metaClassName())
            }
        }
    }

}